src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,273 @@
+            if (tokens.Count == 0)
+            {
+                // We couldn't find any tokens associated with the passed-in range
+                cachedTokens = null;
+                return false;
+            }
+
+            Assumes.NotNull(cachedRangeStart);
+
+            // We can potentially return a partial range match if we can't find a full range match.
+            var endPosition = new Position { Line = cachedRangeStart.Line + cachedRangeLength, Character = 0 };
+            var range = new Range { Start = cachedRangeStart, End = endPosition };
+
+            cachedTokens = (range, tokens);
+            return true;
+        }
+
+        private void CacheTokensPerLine(Range range, int[] tokens, Dictionary<int, IReadOnlyList<int>> lineToTokensDict)
+        {
+            var absoluteLine = 0;
+
+            // Cache the tokens associated with each line
+            var tokenIndex = 0;
+            while (tokenIndex < tokens.Length)
+            {
+                absoluteLine += tokens[tokenIndex];
+
+                // Don't cache tokens outside of the given range
+                if (absoluteLine < range.Start.Line)
+                {
+                    tokenIndex += TokenSize;
+                    continue;
+                }
+
+                if (absoluteLine > range.End.Line)
+                {
+                    break;
+                }
+
+                lock (_dictLock)
+                {
+                    if (lineToTokensDict.TryGetValue(absoluteLine, out _))
+                    {
+                        // This line is already cached, so we can skip it
+                        tokenIndex += TokenSize;
+                        continue;
+                    }
+                }
+
+                // Cache until we hit the next line or the end of the tokens
+                var lineTokens = new List<int>();
+                do
+                {
+                    lineTokens.Add(tokens[tokenIndex]);
+                    lineTokens.Add(tokens[tokenIndex + 1]);
+                    lineTokens.Add(tokens[tokenIndex + 2]);
+                    lineTokens.Add(tokens[tokenIndex + 3]);
+                    lineTokens.Add(tokens[tokenIndex + 4]);
+
+                    tokenIndex += 5;
+                }
+                while (tokenIndex < tokens.Length && tokens[tokenIndex] == 0);
+
+                lock (_dictLock)
+                {
+                    lineToTokensDict.Add(absoluteLine, lineTokens);
+                }
+            }
+        }
+
+        private void FillInEmptyLineGaps(Range range, Dictionary<int, IReadOnlyList<int>> lineToTokensDict)
+        {
+            // Fill in the gaps for empty lines. When looking up tokens later, this helps us differentiate
+            // between empty lines vs. lines we don't have cached tokens for.
+            for (var lineIndex = range.Start.Line; lineIndex < range.End.Line; lineIndex++)
+            {
+                lock (_dictLock)
```

Comment:
taking lock in tight look not recommended.  can you take this lock outside the loop and then just go and set all these empty arrays in the dic.